### PR TITLE
[Inductor] Fallback scatter when src dtype is bf16

### PR DESCRIPTION
**Pull request by oulgen**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113204

basic_gnn_gcn, basic_gnn_gin, basic_gnn_sage now pass

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -3079,6 +3079,8 @@ def scatter_fallback(
     reduce_ty = "add" if fn == "aten.scatter_" else "sum"
     if (
         reduce not in {None, reduce_ty}
+        # tl.atomic_add does not support bf16
+        or src.get_dtype() in {torch.bfloat16}
```
Comment (oulgen) [inline torch/_inductor/lowering.py:3083]: leaving this as a set since i assume there will need to be more things here

Comment (eellison) [reply torch/_inductor/lowering.py:3083]: use tuple? for single elem set seems like overkill

**Review by eellison — COMMENTED** [review]
Can you add a fallback for index_put from this issue while we're at it:

see: https://github.com/pytorch/pytorch/issues/97016

add test ?

**Review by Chillee — CHANGES_REQUESTED** [review]
I think we need to guard this on what GPU we're running on? I think `tl.atomic_add` only doesn't work on A100 GPUs and below.

**Comment by oulgen** [comment]
@Chillee How can i check which gpu i am currently on? or rather how to say A100 or below?

**Comment by oulgen** [comment]
@eellison do you want me to combine both of those tl.atomic_add does not work checks?

**Comment by eellison** [comment]
@Chillee as far as I can tell it's not supported in triton regardless of the device: https://github.com/openai/triton/issues/1387

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -2950,8 +2955,7 @@ def index_put_impl_(self, indices, values, accumulate, check):
     x_size = self.get_size()
     x_ndim = len(x_size)
 
-    # fallback to aten.index_put_, as tl.atomic_add does NOT support int64 or bool
-    if self.get_dtype() in {torch.int64, torch.bool}:
+    if needs_fallback_due_to_atomic_add_limitations(self.get_dtype()):
```
Comment (eellison) [inline torch/_inductor/lowering.py:2958]: `self` can be float, see test failures:

FAILED [0.0248s] inductor/test_torchinductor.py::CpuTests::test_scatter3_cpu - torch._dynamo.exc.BackendCompilerFailed: backend='compile_fx_wrapper' raised:
LoweringException: AttributeError: 'float' object has no attribute 'get_dtype'
  target: aten.scatter.value_reduce

Comment (oulgen) [reply torch/_inductor/lowering.py:2958]: if self can be float, how did this code even work before? weren't we previously also unconditionally calling sef.get_dtype()?

Comment (eellison) [reply torch/_inductor/lowering.py:2958]: Sorry, it may be the other callsite. Anyway check the failing tests

Comment (oulgen) [reply torch/_inductor/lowering.py:2958]: yep, already fixed that one, i thought there was something i missed here

**Review by eellison — APPROVED** [review]
looks good, thanks!

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -2926,6 +2926,11 @@ def _unsafe_index_put_(self, indices, values, accumulate=False):
     return index_put_impl_(self, indices, values, accumulate, check=False)
 
 
+def needs_fallback_due_to_atomic_add_limitations(dtype):
+    # tl.atomic_add does NOT support the following types
+    return dtype in {torch.int64, torch.bool, torch.bfloat16}
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:2931]: Shouldn't this check the device as well?

Comment (oulgen) [reply torch/_inductor/lowering.py:2931]: Yeah, i guess we should only do the pessimization in cuda mode and not cpu mode.

**Comment by jansel** [comment]
cc @htyu

**Pull request closed without merging**